File: tests/.hooks/go/custom-http/internal/hooks/customsecurityhook.go
package hooks

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"

	"github.com/speakeasy/customhttp/models/components"
)

type CustomSecurityHook struct{}

var _ beforeRequestHook = (*CustomSecurityHook)(nil)

func (i *CustomSecurityHook) BeforeRequest(hookCtx BeforeRequestContext, req *http.Request) (*http.Request, error) {
	switch hookCtx.OperationID {
	case "customHttpOnly":
		if hookCtx.SecuritySource == nil {
			return nil, errors.New("security source is nil")
		}
		sec, err := hookCtx.SecuritySource(req.Context())
		if err != nil {
			return nil, err
		}

		security, ok := sec.(*components.Security)
		if !ok {
			return nil, errors.New("security source is not of type Security")
		}

		if customHTTP := security.GetCustomHTTP(); customHTTP == nil {
			return nil, errors.New("customHTTP security is not defined")
		} else {
			req.Header.Set("X-Security-UserID", fmt.Sprintf("%d", customHTTP.GetUserID()))
			req.Header.Set("X-Security-Role", string(customHTTP.GetRole()))
			req.Header.Set("X-Security-Passphrase", customHTTP.GetPassphrase())

			if accessCode := customHTTP.GetAccessCode(); accessCode != nil {
				req.Header.Set("X-Security-AccessCode", fmt.Sprintf("%d", *accessCode))
			}

			if scopes := customHTTP.GetScopes(); len(scopes) > 0 {
				scopesJSON, err := json.Marshal(scopes)
				if err != nil {
					return nil, fmt.Errorf("failed to marshal scopes: %w", err)
				}
				req.Header.Set("X-Security-Scopes", string(scopesJSON))
			}
		}
	}

	return req, nil
}

File: tests/.hooks/pythonv2/custom-http/src/customhttp/_hooks/customsecurityhook.py
from typing import Union, cast
import httpx
import json

from customhttp.models.security import Security

from customhttp._hooks.types import (
    BeforeRequestContext,
    BeforeRequestHook,
)


class CustomSecurityHook(BeforeRequestHook):
    def before_request(
        self, hook_ctx: BeforeRequestContext, request: httpx.Request
    ) -> Union[httpx.Request, Exception]:
        if hook_ctx.operation_id == "customHttpOnly":
            if hook_ctx.security_source is None:
                return Exception("security source is None")

            sec = hook_ctx.security_source
            if callable(sec):
                sec = sec()

            security = cast(Security, sec)
            custom_http = security.custom_http
            if custom_http is None:
                return Exception("custom_http is not defined")

            request.headers["X-Security-UserID"] = str(custom_http.user_id)
            request.headers["X-Security-Role"] = custom_http.role
            request.headers["X-Security-Passphrase"] = custom_http.passphrase

            if custom_http.access_code is not None:
                request.headers["X-Security-AccessCode"] = str(custom_http.access_code)

            if custom_http.scopes is not None and len(custom_http.scopes) > 0:
                request.headers["X-Security-Scopes"] = json.dumps(custom_http.scopes)

        return request
